yt_dlp/output/progress.py

```diff
@@ -0,0 +1,158 @@
+        self.output = output
+        self.maximum = lines - 1
+        self.preserve = preserve
+        self.newline = newline
+        self._lastline = 0
+        self._lastlength = 0
+
+    @_synchronized
+    def print_at_line(self, text, pos: int):
+        if not self.output:
+            return
+
+        if self.newline:
+            self._write(self._add_line_number(text, pos), '\n')
+            return
+
+        if self.output.use_color:
+            self._write(self._move_to(pos), ERASE_LINE, text)
+            return
+
+        text = self._add_line_number(text, pos)
+        textlen = len(text)
+        if self._lastline == pos:
+            # move cursor at the start of progress when writing to same line
+            prefix = '\r'
+            if self._lastlength > textlen:
+                text += ' ' * (self._lastlength - textlen)
+        else:
+            # otherwise, break the line
+            prefix = '\n'
+
+        self._write(prefix, text)
+        self._lastlength = textlen
+        self._lastline = pos
+
+    @_synchronized
+    def close(self):
+        if not self.output or self.newline:
+            return
+
+        # move cursor to the end of the last line and write line break
+        if self.preserve:
+            if self.output.use_color:
+                self._write(self._move_to(self.maximum), '\n')
+            else:
+                self._write('\n')
+            return
+
+        # Try to clear as many lines as possible
+        if self.output.use_color:
+            self._write(
+                self._move_to(self.maximum), ERASE_LINE,
+                f'{MOVE_UP}{ERASE_LINE}' * self.maximum)
+        else:
+            self._write('\r', ' ' * self._lastlength, '\r')
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def _write(self, *text):
+        self.output.log(''.join(text))
+
+    def _add_line_number(self, text, line):
+        if self.maximum:
+            return f'{line + 1}: {text}'
+        return text
+
+    def _move_to(self, line):
+        distance = line - self._lastline
+        self._lastline = line
+        return f'\r{move_cursor(distance)}'
+
+
+class ProgressReporter:
+    def __init__(self, ydl, prefix, *, lines=1, newline=False, preserve=False, disabled=False, templates={}):
+        self._ydl = ydl
+        self._progress = Progress.make_progress(ydl.logger, lines=lines, preserve=preserve, newline=newline)
+        self._console = ydl.console
+        self.disabled = disabled
+        # Pass in a `progress_template` (`params.get('progress_template', {})`)
+        self._screen_template = templates.get(prefix) or f'[{prefix}] %(progress._default_template)s'
+        self._title_template = templates.get(f'{prefix}-title') or 'yt-dlp %(progress._default_template)s'
+        self._finish_template = templates.get(f'{prefix}-finish') or f'[{prefix}] {prefix.capitalize()} completed'
```

Comment:
If this ends up getting used by postprocessor (or others for that matter), `postprocessor-title` and `postprocessor-finish` (`other-title`, ...) would become supported. Is this good or bad?